### PR TITLE
[SPARK-48350][SQL] Introduction of Custom Exceptions for Sql Scripting

### PR DESCRIPTION
**Pull request by miland-db**
### What changes were proposed in this pull request?
Previous PRs introduced basic changes for SQL Scripting. This PR is a follow-up to introduce custom exceptions that can arise while using SQL Scripting language.

### Why are the changes needed?
The intent is to add precise errors for various SQL scripting concepts.

### Does this PR introduce any user-facing change?
Users will now see specific SQL Scripting language errors.

### How was this patch tested?
There are tests for newly introduced parser changes:

SqlScriptingParserSuite - unit tests for execution nodes.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -142,27 +142,29 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     CompoundBody(buff.toSeq, label)
   }
 
-  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody = {
-    val beginLabelCtx = Option(ctx.beginLabel())
-    val endLabelCtx = Option(ctx.endLabel())
+  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody =
+    withOrigin(ctx) {
```
Comment (davidm-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:146]: we are not using origin anywhere in this function, maybe remove for now and we can add if needed later.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 2]
```diff
@@ -3731,6 +3731,24 @@
     ],
     "sqlState" : "42K0I"
   },
+  "SQL_SCRIPTING_LABEL_ERROR" : {
```
Comment (davidm-db) [inline common/utils/src/main/resources/error/error-conditions.json:3734]: not sure if it's important or not, but maybe we can check with Serge if we want to use `SQL_SCRIPTING_*` for error classes, or maybe stick to the simple `SQL_*` or maybe something else.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/exceptions/SqlScriptingException.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.exceptions
+
+import org.apache.spark.{SparkThrowable, SparkThrowableHelper}
+import org.apache.spark.sql.catalyst.trees.{CurrentOrigin, Origin, WithOrigin}
+
+class SqlScriptingException protected (
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/exceptions/SqlScriptingException.scala:23]: I'm fine with sql scripting specific error conditions, but do we really need a dedicated java exception class? Can we simply use `SparkException`? cc @srielau

**Review by srielau — APPROVED** [review]
Approved with comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/exceptions/SqlScriptingException.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/exceptions/SqlScriptingException.scala:23]: yea, let's remove it then

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/SqlScriptingParserSuite.scala [hunk 4]
```diff
@@ -203,11 +204,11 @@ class SqlScriptingParserSuite extends SparkFunSuite with SQLHelper {
...
         |  SELECT * FROM T;
         |END lbl_end""".stripMargin
-    val e = intercept[SparkException] {
+    val e = intercept[SqlScriptingException] {
       parseScript(sqlScriptText)
     }
-    assert(e.getErrorClass === "INTERNAL_ERROR")
-    assert(e.getMessage.contains("Both labels should be same."))
+    assert(e.getErrorClass === "SQL_SCRIPTING_LABEL_ERROR.LABELS_MISMATCH")
+    assert(e.getMessage.contains("Begin label lbl_begin does not match the end label lbl_end."))
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/SqlScriptingParserSuite.scala:211]: +1

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**